### PR TITLE
Optimize counting digits in line numbers during error reporting

### PR DESCRIPTION
**Pull request by nhwn**
Replaces `.to_string().len()` with simple loop and integer division, which avoids an unnecessary allocation. 

Although I couldn't figure out how to directly profile `rustc`'s error reporting, I ran a microbenchmark on my machine (2.9 GHz Dual-Core Intel Core i5) on the two strategies for `0..100_000`, and the results seem promising:
```
test to_string_len ... bench:  12,124,792 ns/iter (+/- 700,652)
test while_loop    ... bench:      30,333 ns/iter (+/- 562)
```
The x86_64 disassembly reduces integer division to a multiplication + shift, so I don't think there's any problems with using integer division.

For more (micro)optimization, it would be nice if we could avoid the initial check to see if the line number is nonzero, but I don't think `self.get_max_line_num(span, children)` _guarantees_ a nonzero line number.

### CODE REVIEW

**Comment by leonardo-m** [comment]
An alternative version:

    let mut num_digits = 0;
    loop {
        num_digits += 1;
        n /= 10;
        if n == 0 { break; }
    }
    num_digits

**Comment by nhwn** [comment]
> An alternative version:
> 
> ```
> let mut num_digits = 0;
> loop {
>     num_digits += 1;
>     n /= 10;
>     if n == 0 { break; }
> }
> num_digits
> ```

That's a much better implementation than the original one, thanks!

File: compiler/rustc_errors/src/emitter.rs [hunk 1]
```diff
@@ -1713,7 +1713,15 @@ impl EmitterWriter {
         let max_line_num_len = if self.ui_testing {
             ANONYMIZED_LINE_NUM.len()
         } else {
-            self.get_max_line_num(span, children).to_string().len()
+            let mut n = self.get_max_line_num(span, children);
```
Comment (varkor) [inline compiler/rustc_errors/src/emitter.rs:1716]: Could you add a comment explaining why we calculate the number of digits this way? (To make sure someone doesn't revert this change without realising it's intentional, later.)

**Comment by varkor** [comment]
Could you squash the commits? Thanks.

File: compiler/rustc_errors/src/emitter.rs (hunk 1 above)
Comment (nhwn) [reply compiler/rustc_errors/src/emitter.rs:1719]: Added an explanatory comment.

**Comment by varkor** [comment]
Thanks!

@bors r+ rollup

**Comment by nhwn** [comment]
It seems my `tidy` hook didn't auto-run when I rebased, so a pesky space slipped through in the comment. Sorry about that.

**Comment by matthieu-m** [comment]
I actually just realized there's doesn't appear to be a fast way to compute the number of (decimal) digits that an integer would take.

I find it surprising, as I would have expected integer formatting to make use of this. For reference, a simple version is to switch on the number of bits ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f1d40f97a83037ea2b69847cba41249)):

    fn count_digits(i: u32) -> u32 {
        if i < 10 {
            return 1;
        }
        if i < 100 {
            return 2;
        }
    
        match 32 - i.leading_zeros() {
            0 | 1 | 2 | 3 | 4 | 5 | 6
                //  100 takes 7 bits, and anything < 100 is already handled.
                => unsafe { hint::unreachable_unchecked() },
            7 | 8 | 9 | 10
                => 3 + if i < 1_000 { 0 } else { 1 },
            11 | 12 | 13 | 14
                => 4 + if i < 10_000 { 0 } else { 1 },
            15 | 16 | 17
                => 5 + if i < 100_000 { 0 } else { 1 },
            18 | 19 | 20
                => 6 + if i < 1_000_000 { 0 } else { 1 },
            21 | 22 | 23 | 24
                => 7 + if i < 10_000_000 { 0 } else { 1 },
            25 | 26 | 27
                => 8 + if i < 100_000_000 { 0 } else { 1 },
            28 | 29 | 30
                => 9 + if i < 1_000_000_000 { 0 } else { 1 },
            31 | 32 => 10,
            //  There are not more than 32 bits in a 32 bits integer
            _ => unsafe { hint::unreachable_unchecked() },
        }
    }

It may be a useful addition to the standard library.

**Comment by toothbrush7777777** [comment]
> 
> 
> I actually just realized there's doesn't appear to be a fast way to compute the number of (decimal) digits that an integer would take.
> […]

Basically, you want `ceil(log10(n))`.

**Pull request merged**